### ISSUE TITLE
ClassInfoList.loadClasses() should return a Set and not a List

### ISSUE THREAD

**Issue by GedMarc** [classgraph/classgraph#231]
Results from loadClasses() should never contain duplicates, and the type returned is for duplicates.

Please change to Set<> to identify uniqueness and use the optimizations that Set provides over a List

**Comment by johnou** [comment]
linkedhashset to maintain order if that's important?

**Comment by lukehutch** [comment]
Internally ClassInfoList is deduplicated (the constructor requires a set). So the semantics are already what you are looking for. The reason I made it a list was so that the classes could be sorted in name order for consistency. That's not crucial though, I guess. Is it worth breaking the API again to make the set nature more obvious? 

LinkedHashSet returns elements in addition order, so you would have to create two intermediate collections (a set, then a sorted list) to sorry by name.

**Comment by GedMarc** [comment]
Hmmm, I Suppose not, From what I understand there's a few performance degradations with ArrayList where it is of type "Resizable", allows nulls, and is modifyable without implicit equalizers and sorters.

Either way I suppose both are Collections API, but my understanding is that Set's are fully optimized for unique and are non-resizable, going "Tree" you implement a proper not-null structure and gain all the performance benefits from that as well. 

Considering i'm already having a blonde day, erg
 TreeSet Javadoc says, "This implementation provides guaranteed log(n) time cost for the basic
operations (add, remove and contains)." This implies O(N log N) time to
insert N items. Collections.sort Javadoc says, "This algorithm offers guaranteed n log(n) performance."

**Comment by lukehutch** [comment]
Actually ArrayList has O(1) amortized cost for all operations except for insert/remove at arbitrary position. It is a thin veneer around array access, so it's not much slower than just using an array. A TreeSet has very high overhead, because on top of the O(N log N) operation time, it allocates one wrapper object per data item (a bit like a linked list).

In an earlier version of the API, I made all the *List classes immutable, but then in a few cases I needed to build the lists incrementally in ClassGraph code. So I decided that I would simply return sane lists to the user, then assume that if they start to mess with the contents of the lists, they get to keep both pieces (ie. guarantees of uniqueness go out the window). 

I'm open to the idea of switching the API to return a LinkedHashSet, since that would be more semantically true to intent anyway. But that would be a major change to the API. Is that worth it?
